src/drivers/legos/driver.rs

```diff
@@ -0,0 +1,402 @@
+    bad_data_passed: bool,
+}
+
+impl Driver {
+    pub fn new(path: String) -> Result<Self, Box<dyn Error + Send + Sync>> {
+        let fmtpath = path.clone();
+        let path = CString::new(path)?;
+        let api = hidapi::HidApi::new()?;
+        let device = api.open_path(&path)?;
+        let info = device.get_device_info()?;
+        if info.vendor_id() != VID || info.product_id() != PID {
+            return Err(format!("Device '{fmtpath}' is not a Legion Go S Controller").into());
+        }
+
+        Ok(Self {
+            device,
+            accel_state: None,
+            gyro_state: None,
+            xinput_state: None,
+            bad_data_passed: false,
+        })
+    }
+
+    /// Poll the device and read input reports
+    pub fn poll(&mut self) -> Result<Vec<Event>, Box<dyn Error + Send + Sync>> {
+        // Read data from the device into a buffer
+        let mut buf = [0; XINPUT_PACKET_SIZE];
+        let bytes_read = self.device.read_timeout(&mut buf[..], HID_TIMEOUT)?;
+
+        let events = match bytes_read {
+            XINPUT_PACKET_SIZE => {
+                let events = match self.handle_xinput_report(buf) {
+                    Ok(events) => events,
+                    Err(e) => {
+                        log::error!("Got error processing XinputDataReport: {e:?}");
+                        vec![]
+                    }
+                };
+                events
+            }
+            INERTIAL_PACKET_SIZE => {
+                let slice = &buf[..bytes_read];
+                // Handle the incoming input report
+                let sized_buf = slice.try_into()?;
+                let events = match self.handle_inertial_report(sized_buf) {
+                    Ok(events) => events,
+                    Err(e) => {
+                        log::error!("Got error processing InertailInputDataReport: {e:?}");
+                        vec![]
+                    }
+                };
+                events
+            }
+
+            _ => vec![],
+        };
+
+        Ok(events)
+    }
+
+    /// Unpacks the buffer into a [XinputDataReport] structure and updates
+    /// the internal xinput_state
+    fn handle_xinput_report(
+        &mut self,
+        buf: [u8; XINPUT_PACKET_SIZE],
+    ) -> Result<Vec<Event>, Box<dyn Error + Send + Sync>> {
+        let input_report = XInputDataReport::unpack(&buf)?;
+
+        // Hacky workaround. When the gyro device is grabbed the XInputDataReport is full of
+        // garbage. Since it doesn't have a report_id we cant reject it. This only seems to hapen
+        // one time, so we can save a lot of checks in the future if we clear a bool once it
+        // happens.
+        if !self.bad_data_passed {
+            if input_report.is_bad_data() {
+                log::debug!("Got bad XInputDataReport, regecting it.");
+                self.bad_data_passed = true;
+                return Ok(vec![]);
+            }
+        }
```

Comment:
If the device doesn't support report ids, this might be the best thing we can do, unless the vendor can provide more information.

To fix the clippy warning, though:

```suggestion
        // Hacky workaround. When the gyro device is grabbed the XInputDataReport is full of
        // garbage. Since it doesn't have a report_id we cant reject it. This only seems to hapen
        // one time, so we can save a lot of checks in the future if we clear a bool once it
        // happens.
        if !self.bad_data_passed && input_report.is_bad_data() {
            log::debug!("Got bad XInputDataReport, regecting it.");
            self.bad_data_passed = true;
            return Ok(vec![]);
        }
```